FakeTensor speedup: non-kwargs for TensorMetadata

aorenste opened the pull request:
It turns out that using kwargs, even if they're exactly in the same order as the positional args, is significantly slower in hot code.

2.1% FakeTensor perf win on the microbenchmark (25.2% cumulative)


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125312
* __->__ #124226
* #124225

aorenste committed:
non-kwargs for TensorMetadata

[ghstack-poisoned]

aorenste committed:
Update on "non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "FakeTensor speedup: non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "FakeTensor speedup: non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "FakeTensor speedup: non-kwargs for TensorMetadata"




[ghstack-poisoned]

aorenste committed:
Update on "FakeTensor speedup: non-kwargs for TensorMetadata"




[ghstack-poisoned]

masnesral reviewed (approved):
wut?! I wouldn't have guessed.

oulgen reviewed (approved):
this is just messed up..

oulgen commented:
@eellison For statically known function calls, I would have guessed the runtime is able to elide the extra cost but I guess this might be more the reason why we would be using a compiled language?

aorenste committed:
Update on "FakeTensor speedup: non-kwargs for TensorMetadata"

It turns out that using kwargs, even if they're exactly in the same order as the positional args, is significantly slower in hot code.

2.1% FakeTensor perf win on the microbenchmark (25.2% cumulative)





[ghstack-poisoned]

aorenste committed:
Update on "FakeTensor speedup: non-kwargs for TensorMetadata"

It turns out that using kwargs, even if they're exactly in the same order as the positional args, is significantly slower in hot code.

2.1% FakeTensor perf win on the microbenchmark (25.2% cumulative)





[ghstack-poisoned]
